presto-raptor/src/main/java/com/facebook/presto/raptor/storage/BucketBalancer.java

```diff
@@ -0,0 +1,369 @@
+
+            for (String source : ImmutableSet.copyOf(allocationCounts)) {
+                List<BucketAssignment> existingAssignments = distributionAssignments.stream()
+                        .filter(assignment -> assignment.getNodeIdentifier().equals(source))
+                        .collect(toList());
+
+                for (BucketAssignment existingAssignment : existingAssignments) {
+                    if (activeNodes.contains(source) && allocationCounts.count(source) <= targetMin) {
+                        break;
+                    }
+
+                    // identify nodes with bucket counts lower than the computed target, and greedily select from this set based on projected disk utilization.
+                    // greediness means that this may produce decidedly non-optimal results if one looks at the global distribution of buckets->nodes.
+                    // also, this assumes that nodes in a cluster have identical storage capacity
+                    String target = activeNodes.stream()
+                            .filter(candidate -> !candidate.equals(source) && allocationCounts.count(candidate) < targetMax)
+                            .sorted((a, b) -> Integer.compare(allocationCounts.count(a), allocationCounts.count(b)))
+                            .min((a, b) -> Double.compare(allocationBytes.get(a), allocationBytes.get(b)))
+                            .orElseThrow(() -> new RuntimeException("unable to find target for rebalancing"));
+
+                    long bucketSize = clusterState.getDistributionBucketSize().get(distribution);
+
+                    // don't move bucket if it doesn't reduce imbalance
+                    if (activeNodes.contains(source) && (allocationCounts.count(source) == targetMax && allocationCounts.count(target) == targetMin)) {
+                        break;
+                    }
+
+                    allocationCounts.remove(source);
+                    allocationCounts.add(target);
+                    allocationBytes.compute(source, (k, v) -> v - bucketSize);
+                    allocationBytes.compute(target, (k, v) -> v + bucketSize);
+
+                    sourceToAllocationChanges.put(
+                            existingAssignment.getNodeIdentifier(),
+                            new BucketAssignment(existingAssignment.getDistributionId(), existingAssignment.getBucketNumber(), target));
+                }
+            }
+        }
+
+        return sourceToAllocationChanges;
+    }
+
+    private int updateAssignments(Multimap<String, BucketAssignment> sourceToAllocationChanges)
+    {
+        // perform distributionAssignments in decreasing order of source node load
+        List<String> sourceNodes = sourceToAllocationChanges.asMap().entrySet().stream()
+                .sorted((a, b) -> Integer.compare(b.getValue().size(), a.getValue().size()))
```

Comment:
Using comparingInt here is ugly, since I need to add a cast to make type inference work properly with `reversed()`